README.md
# HTCC
[![Gem Version](https://badge.fury.io/rb/htcc.svg)](https://badge.fury.io/rb/htcc)

This gem is a Ruby client for the Honeywell Total Connect Comfort API.

I have only tested it with the Honeywell Home Wi-Fi 7-Day Programmable Thermostat (RTH6580WF), but it should work with any thermostats that use Total Connect Comfort for remote control. I would be interested in knowing what other devices have for `thermostat.info['DeviceType']`. I'm currently initializing everything as an instance of `HTCC::Thermostat`.

Currently, settings and scheduling are a work in progress.

## Installation
`gem install htcc`

You can also test it out by forking this respository and running `docker-compose run htcc`.
## Basic Usage
```ruby
htcc = HTCC::Client.new('tcc_login@example.com', 'password')
# => #<HTCC::Client:0x0000abc123 @debug=false...>
```

Get first device.
```ruby
thermostat = htcc.devices.first
# => #<HTCC::Thermostat:0x0000abc123 @info={"DeviceID"=>...>
```

Get the current ambient temperature.
```ruby
thermostat.current_temperature
# => 71
```

Get current setpoint for heat.
```ruby
thermostat.heat_setpoint
# => 70
```

Set the heat setpoint to desired temperature with a temporary hold. Returns to scheduled temperature at the next scheduled time.
```ruby
thermostat.heat_setpoint = 72
# => 72
```

## `HTCC::Client` Usage
```ruby
htcc = HTCC::Client.new('tcc_login@example.com', 'password', debug: true, debug_output: $stdout)
# => #<HTCC::Client:0x0000abc123 @debug=true...>
```
`debug` determines whether to display the HTTP request/response data (boolean, default: `false`)

>__Warning!__ If `debug` is set to `true` when initializing the `HTCC::Client` instance, it will display the login and password in plain text. You can also set `debug` to `true` *after* initializing the client. However, this will also display sensitive data such as cookies for every request. See `Net::HTTP` documentation. __Use at your own risk__.

`debug_output` sets the output stream for debugging (Used by `Net::HTTP#set_debug_output`, default: `$stdout`).

Display HTTP request/response data.
```ruby
htcc.debug = true
# => true
```
> __See above warning.__

Check if log in was successful.
```ruby
htcc.logged_in?
# => true
```

Get list of devices associated with logged in account Accepts an optional boolean argument to refresh the device list (default: `false`).
```ruby
htcc.devices
# => [#<HTCC::Thermostat:0x0000abc123...>]

htcc.devices(true)
# => [#<HTCC::Thermostat:0x0000abc123...>]
```

## `HTCC::Thermostat` Usage
Get information about the thermostat.
```ruby
thermostat.info
# => {"DeviceID=>1234567, "DeviceType"=>24, "LocationID=>0123456, "Name"=>"THERMOSTAT", "IsAlive"=>true...}
```

Get device ID.
```ruby
thermostat.id
# => 1234567
```

Get device MAC address.
```ruby
thermostat.mac_address
# => "00A00B0C00D000"
```

Get device name.
```ruby
thermostat.name
# => "THERMOSTAT"
```

Get connection status.
```ruby
thermostat.connected?
# => true
```

Get the status of the device. Accepts an optional boolean argument to refresh the status (default: `false`).
```ruby
thermostat.status
# => {"success"=>true, "deviceLive"=>true, "communicationLost"=>false, "latestData"=>...}

thermostat.status(true)
# => {"success"=>true, "deviceLive"=>true, "communicationLost"=>false, "latestData"=>...}
```

Get current system mode. Returns `:emergency_heat`, `:heat`, `:off`, `:cool`, or `:auto` depending on the capabilities of your device.
```ruby
thermostat.system_mode
# => :heat
```

Set system mode. Returns an `HTCC::Thermostat::SystemError` if desired mode is not present.
```ruby
thermostat.system_mode = :cool
# => :cool

thermostat.system_mode = :emergency_heat
# => HTCC::Thermostat::SystemError (Unknown mode: :emergency_heat. Allowed modes: [:heat, :off, :cool, :auto])
```

Check if the device has a fan.
```ruby
thermostat.has_fan?
# => true
```

Check if the fan is running.
```ruby
thermostat.fan_running?
# => false
```

Get the fan mode. Returns `:auto`,  `:on`,  `:circulate`, or `:schedule` depending on the capabilities of your device.
```ruby
thermostat.fan_mode
# => :auto
```

Set the fan mode. Returns an `HTCC::Thermostat::FanError` if desired mode is not present.
```ruby
thermostat.fan_mode = :on
# => :on

thermostat.fan_mode = :circulate
# => HTCC::Thermostat::FanError (Unknown mode: :circulate. Allowed modes: [:auto, :on])
```

Get the current ambient temperature.
```ruby
thermostat.current_temperature
# => 72
```

Get the current temperature unit ("C" or "F").
```ruby
thermostat.temperature_unit
# => "F"
```

Get the temperature that is set for cooling.
```ruby
thermostat.cool_setpoint
# => 68
```

Set the temperature for cooling. Uses a temporary hold that returns to scheduled temperature at the next scheduled time. Returns an `HTCC:Thermostat::TemperatureError` if the given temperature is out of range.
```ruby
thermostat.cool_setpoint = 70
# => 70

thermostat.cool_setpoint = 32
# => HTCC::Thermostat::TemperatureError (Minimum setpoint is 50. Given: 32)
```

Get the cool setpoint range.
```ruby
thermostat.cool_setpoint_range
# => 50..99
```

Get the temperature that is set for heating.
```ruby
thermostat.heat_setpoint
# => 72
```

Set the temperature for heating. Uses a temporary hold that returns to scheduled temperature at the next scheduled time. Returns an `HTCC:Thermostat::TemperatureError` if the given temperature is out of range.
```ruby
thermostat.heat_setpoint = 74
# => 74

thermostat.heat_setpoint = 100
# => HTCC::Thermostat::TemperatureError (Maximum setpoint is 90. Given: 100)
```

Get the heat setpoint range.
```ruby
thermostat.heat_setpoint_range
# => 40..90
```

Get the current hold status. Returns `:none`, `:temporary`, or `:permanent` depending on the capabilities of your device.
```ruby
thermostat.hold
# => :none
```

Set a hold. Returns an `HTCC::Thermostat::HoldError` if desired mode is not present.
```ruby
thermostat.hold = :temporary
# => :temporary

thermostat.hold = :permanent
# => HTCC::Thermostat::HoldError (Unknown mode: :permanent. Allowed modes: [:none, :temporary])
```

Get the time a temporary hold ends. Returns a 24-hour formatted string (HH:MM). If the current hold is `:permanent`, this time is ignored unless the hold is changed to `:temporary`.
```ruby
thermostat.hold_until
# => "22:15"
```

Set a temporary hold until a given time. Time must be in a zero-padded 24-hour format (HH:MM). Time begin at 00:00 (midnight) to 23:45. Minutes must be in quarter hours (:00, :15, :30, :45). Returns an `HTCC::Thermostat::HoldError` if given time is not valid.
```ruby
thermostat.hold_until = '05:45'
# => "05:45"

thermostat.hold_until = '05:46'
# => HTCC::Thermostat::HoldError (Unknown hold time: "05:46". Valid times are from 00:00 - 23:45 in 15 minute intervals.)
```

Clear out any holds and resume heating/cooling schedule (same as `thermostat.hold = :none`).
```ruby
thermostat.resume_schedule
# => true
```

Get the current output status. Returns `:off`, `:heating`, `:cooling`, or `:fan_running` depending on the capabilities of your device.
```ruby
thermostat.output_status
# => :cooling
```

Get the available system modes.
```ruby
thermostat.system_modes
# => [:heat, :off, :cool, :auto]
```

Get the available fan modes.
```ruby
thermostat.fan_modes
# => [:auto, :on]
```

Call a property without making an API request to the TCC API (using the in-memory data).
```ruby
thermostat.no_refresh { thermostat.current_temperature }
# => 78
```

## `HTCC::Scheduler` Usage
TODO

## `HTCC::Settings` Usage
TODO

CHANGELOG.md
# Changelog

## 0.3.1 (29-March-2021)
* __New Features:__
  * Add `#hold_until` and `#hold_until=` to `HTCC::Thermostat` for getting/setting temporary holds until a given time.

## 0.3.0 (29-March-2021)
* __Breaking changes:__
  * Change `#cool_setpoint_range` and `#heat_setpoint_range` on `HTCC::Thermostat` to return a `Range` object for more flexibility (can use `min`, `max`, and `include?` now).

## 0.2.0 (29-March-2021)
* __Breaking changes:__
  * Removed `#refresh_devices` method from `HTCC::Client` in favor of `#devices(true)` for consistency with `HTCC::Thermostat#status` method.

* __New Features:__
  * Add `#cool_setpoint_range` and `#heat_setpoint_range` to `HTCC::Thermostat` for obtaining heating/cooling setpoint limits.
  * Add documentation regarding limits for heat/cool setpoints.

* __Bug Fixes:__
  * Fix README typos.

## 0.1.1 (28-March-2021)

* __Bug Fixes:__
  * Private `HTCC::Thermostat` method `#get_status` should check if `@status` is `empty?`

## 0.1.0 (28-March-2021)

* Initial release


lib/htcc/client.rb
# frozen_string_literal: true

require 'net/http'
require 'json'

module HTCC
  class Client
    BASE_URL = 'https://mytotalconnectcomfort.com/portal'.freeze
    HEADERS  = { 'X-Requested-With': 'XMLHttpRequest' }.freeze

    def initialize(username, password, debug: false, debug_output: nil)
      @debug = debug
      @debug_output = nil
      @devices = []
      login(username, password)
    end

    def debug=(val)
      @debug = val
    end

    def logged_in?
      @logged_in
    end

    def devices(refresh = false)
      return @devices unless @devices.empty? || refresh

      @devices = get_devices if logged_in?
    end

    private

    def get_devices
      resp = request('/Location/GetLocationListData', method: 'post', data: { 'page': '1', 'filter': '' })
      locations = ::JSON.parse(resp.body)
      locations.flat_map { |loc| loc['Devices'] }.map do |device|
        case device['DeviceType']
        when 24
          Thermostat.new(device, self)
        else # Other devices?
          Thermostat.new(device, self)
        end
      end
    end

    def login(username, password)
      resp = request(method: 'post', data: {
        'UserName': username, 'Password': password, 'timeOffset': '240', 'RememberMe': 'false'
      })
      @cookies = get_cookies(resp)
      @logged_in = resp.get_fields('content-length')[0].to_i < 25 # Successful login
    end

    def get_cookies(response)
      response.get_fields('set-cookie')
              .map { |c| cookie = c.split(/;|,/)[0]; cookie.split('=')[1] ? cookie : nil }
              .compact
              .join(';')
    end

    def request(path = '', method: 'get', data: nil, headers: {})
      uri = URI("#{BASE_URL}#{path}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      http.set_debug_output(@debug_output || $stdout) if @debug
      klass = method == 'get' ? Net::HTTP::Get : Net::HTTP::Post
      request = klass.new(uri.request_uri, HEADERS.merge(headers))
      request['Cookie'] = @cookies if @cookies
      request.set_form_data(data) if data
      http.request(request)
    end
  end
end


lib/htcc/settings.rb
# frozen_string_literal: true

module HTCC
  class Settings
    attr_reader :device_id

    def initialize(device_id, client)
      @device_id = device_id
      @client = client
    end

    def update(payload)
      resp = @client.send(:request, '/Device/Menu/Settings', method: 'post', data: payload)
      JSON.parse(resp.body)
    end
  end
end

# TODO:

# Payload that can be posted to '/Device/Menu/Settings'
# {
#   "Name":"THERMOSTAT",
#   "ApplySettingsToAllZones":false,
#   "DeviceID":123456,
#   "DisplayUnits":1,
#   "TempHigherThanActive":true,
#   "TempHigherThan":85,
#   "TempHigherThanMinutes":15,
#   "TempLowerThanActive":true,
#   "TempLowerThan":55,
#   "TempLowerThanMinutes":15,
#   "HumidityHigherThanActive":null,
#   "HumidityHigherThan":null,
#   "HumidityHigherThanMinutes":null,
#   "HumidityLowerThanActive":null,
#   "HumidityLowerThan":null,
#   "HumidityLowerThanMinutes":null,
#   "FaultConditionExistsActive":false,
#   "FaultConditionExistsHours":1,
#   "NormalConditionsActive":true,
#   "ThermostatAlertActive":null,
#   "CommunicationFailureActive":true,
#   "CommunicationFailureMinutes":15,
#   "CommunicationLostActive":true,
#   "CommunicationLostHours":1,
#   "DeviceLostActive":null,
#   "DeviceLostHours":null,
#   "TempHigherThanValue":"85°",
#   "TempLowerThanValue":"55°",
#   "HumidityHigherThanValue":"--%",
#   "HumidityLowerThanValue":"--%",
#   "TempHigherThanMinutesText":"For 15 Minutes",
#   "TempLowerThanMinutesText":"For 15 Minutes",
#   "HumidityHigherThanMinutesText":"For 0 Minutes",
#   "HumidityLowerThanMinutesText":"For 0 Minutes",
#   "FaultConditionExistsHoursText":"Every 1 Hour",
#   "CommunicationFailureMinutesText":"For 15 Minutes",
#   "CommunicationLostHoursText":"After 1 Hour",
#   "DeviceLostHoursText":"After 1 Hour"
# }


lib/htcc/scheduler.rb
# frozen_string_literal: true

module HTCC
  class Scheduler
    attr_reader :device_id

    def initialize(device_id, client)
      @device_id = device_id
      @client = client
    end

    def get_schedule
      resp = @client.send(:request, "/Device/Menu/GetScheduleData/#{device_id}", method: 'post')
      @schedule = JSON.parse(resp.body)
    end
  end
end

# TODO:

# EDIT_SCHEDULE_PATH = '/Device/Menu/EditSchedule/' + device_id
# CONFIRM_SCHEDULE_PATH = '/Device/Menu/SendSchedule?deviceId=' + device_id
# DISCARD_SCHEDULE_PATH = '/Device/Menu/DiscardChangesInSchedule?deviceID=' + device_id

# FormData Encoded:

# DeviceID: 123456
# Days[0]: True
# Days[1]: True
# Days[2]: True
# Days[3]: True
# Days[4]: True
# Days[5]: True
# Days[6]: True
# DayChange: True
# Templates[0].Editable: True
# Templates[0].OrigIsCancelled: false
# Templates[0].Editable: True
# Templates[0].Type: WakeOcc1
# Templates[0].OrigStartTime: 05:45:00
# Templates[0].OrigFanMode: Auto
# Templates[0].OrigHeatSetpoint: 68
# Templates[0].OrigCoolSetpoint: 79
# Templates[0].StartTime: 05:45:00
# Templates[0].IsCancelled: false
# Templates[1].Editable: True
# Templates[1].OrigIsCancelled: true
# Templates[1].Editable: True
# Templates[1].Type: LeaveUnocc1
# Templates[1].OrigStartTime: 08:00:00
# Templates[1].OrigFanMode: Auto
# Templates[1].OrigHeatSetpoint: 62
# Templates[1].OrigCoolSetpoint: 85
# Templates[1].StartTime: 08:00:00
# Templates[1].IsCancelled: true
# Templates[2].Editable: True
# Templates[2].OrigIsCancelled: true
# Templates[2].Editable: True
# Templates[2].Type: ReturnOcc2
# Templates[2].OrigStartTime: 18:00:00
# Templates[2].OrigFanMode: Auto
# Templates[2].OrigHeatSetpoint: 70
# Templates[2].OrigCoolSetpoint: 78
# Templates[2].StartTime: 18:00:00
# Templates[2].IsCancelled: true
# Templates[3].Editable: True
# Templates[3].OrigIsCancelled: false
# Templates[3].Editable: True
# Templates[3].Type: SleepUnocc2
# Templates[3].OrigStartTime: 21:30:00
# Templates[3].OrigFanMode: Auto
# Templates[3].OrigHeatSetpoint: 64
# Templates[3].OrigCoolSetpoint: 78
# Templates[3].StartTime: 21:30:00
# Templates[3].IsCancelled: false
# Templates[0].Type: WakeOcc1
# Templates[0].HeatSetpoint: 68
# Templates[0].CoolSetpoint: 79
# Templates[1].Type: LeaveUnocc1
# Templates[1].HeatSetpoint: 62
# Templates[1].CoolSetpoint: 85
# Templates[2].Type: ReturnOcc2
# Templates[2].HeatSetpoint: 70
# Templates[2].CoolSetpoint: 78
# Templates[3].Type: SleepUnocc2
# Templates[3].HeatSetpoint: 64
# Templates[3].CoolSetpoint: 78
# DisplayUnits: Fahrenheit
# IsCommercial: False
# ScheduleFan: True
# Templates[0].FanMode: Auto
# Templates[1].FanMode: Auto
# Templates[2].FanMode: Auto
# Templates[3].FanMode: Auto
# ScheduleOtherDays: False


lib/htcc/thermostat.rb
# frozen_string_literal: true

module HTCC
  class Thermostat
    SYSTEM_MODES = %i[emergency_heat heat off cool auto].freeze
    FAN_MODES = %i[auto on circulate schedule].freeze
    HOLD_TYPES = %i[none temporary permanent].freeze
    EQUIPMENT_OUTPUT_STATUS = %i[off heating cooling fan_running].freeze
    HOLD_TIMES = [
      '00:00', '00:15', '00:30', '00:45',
      '01:00', '01:15', '01:30', '01:45',
      '02:00', '02:15', '02:30', '02:45',
      '03:00', '03:15', '03:30', '03:45',
      '04:00', '04:15', '04:30', '04:45',
      '05:00', '05:15', '05:30', '05:45',
      '06:00', '06:15', '06:30', '06:45',
      '07:00', '07:15', '07:30', '07:45',
      '08:00', '08:15', '08:30', '08:45',
      '09:00', '09:15', '09:30', '09:45',
      '10:00', '10:15', '10:30', '10:45',
      '11:00', '11:15', '11:30', '11:45',
      '12:00', '12:15', '12:30', '12:45',
      '13:00', '13:15', '13:30', '13:45',
      '14:00', '14:15', '14:30', '14:45',
      '15:00', '15:15', '15:30', '15:45',
      '16:00', '16:15', '16:30', '16:45',
      '17:00', '17:15', '17:30', '17:45',
      '18:00', '18:15', '18:30', '18:45',
      '19:00', '19:15', '19:30', '19:45',
      '20:00', '20:15', '20:30', '20:45',
      '21:00', '21:15', '21:30', '21:45',
      '22:00', '22:15', '22:30', '22:45',
      '23:00', '23:15', '23:30', '23:45'
    ].freeze

    attr_reader :info

    def initialize(info, client)
      @info = info
      @client = client
      @refresh = true
      @status = {}
    end

    def Scheduler
      @scheduler ||= Scheduler.new(id, @client)
    end

    def Settings
      @settings ||= Settings.new(id, @client)
    end

    def id
      @info['DeviceID']
    end

    def mac_address
      @info['MacID']
    end

    def name
      @info['Name']
    end

    def connected?
      get_status
      @status['deviceLive'] && !@status['communicationLost']
    end

    def status(refresh = false)
      get_status if @status.empty? || refresh
      @status
    end

    def system_mode
      get_status
      SYSTEM_MODES[@status['latestData']['uiData']['SystemSwitchPosition']]
    end

    def system_mode=(mode)
      unless system_modes.index(mode)
        raise SystemError.new("Unknown mode: #{mode.inspect}. Allowed modes: #{system_modes.inspect}")
      end
      change_setting(system_mode: mode)
    end

    def has_fan?
      return @has_fan unless @has_fan.nil?

      get_status if @status.empty?
      @has_fan = @status['latestData']['hasFan']
    end

    def fan_running?
      get_status
      @status['latestData']['fanData']['fanIsRunning']
    end

    def fan_mode
      get_status
      FAN_MODES[@status['latestData']['fanData']['fanMode']]
    end

    def fan_mode=(mode)
      unless fan_modes.index(mode)
        raise FanError.new("Unknown mode: #{mode.inspect}. Allowed modes: #{fan_modes.inspect}")
      end
      change_setting(fan_mode: mode)
    end

    # Current ambient temperature
    def current_temperature
      get_status
      @status['latestData']['uiData']['DispTemperature']
    end

    def temperature_unit
      get_status
      @status['latestData']['uiData']['DisplayUnits']
    end

    # Cooling temperature setting
    def cool_setpoint
      get_status
      @status['latestData']['uiData']['CoolSetpoint']
    end

    def cool_setpoint=(temp)
      raise_min_setpoint(min_cool_setpoint, temp) if temp < min_cool_setpoint
      raise_max_setpoint(max_cool_setpoint, temp) if temp > max_cool_setpoint
      change_setting(cool_setpoint: temp, hold: :temporary)
    end

    def cool_setpoint_range
      (min_cool_setpoint..max_cool_setpoint)
    end

    # Heating temperature setting
    def heat_setpoint
      get_status
      @status['latestData']['uiData']['HeatSetpoint']
    end

    def heat_setpoint=(temp)
      raise_min_setpoint(min_heat_setpoint, temp) if temp < min_heat_setpoint
      raise_max_setpoint(max_heat_setpoint, temp) if temp > max_heat_setpoint
      change_setting(heat_setpoint: temp, hold: :temporary)
    end

    def heat_setpoint_range
      (min_heat_setpoint..max_heat_setpoint)
    end

    def resume_schedule
      change_setting(hold: :none)
    end

    def hold
      get_status
      HOLD_TYPES[@status['latestData']['uiData']['StatusHeat']] # Both statuses return the same value
    end

    def hold=(mode)
      unless HOLD_TYPES.index(mode)
        raise HoldError.new("Unknown mode: #{mode.inspect}. Allowed modes: #{HOLD_TYPES.inspect}")
      end
      change_setting(hold: mode)
    end

    def hold_until
      get_status
      HOLD_TIMES[@status['latestData']['uiData']['HeatNextPeriod']] # Both periods return the same value
    end

    def hold_until=(time)
      unless HOLD_TIMES.index(time)
        raise HoldError.new(
          "Unknown hold time: #{time.inspect}. Valid times are from 00:00 - 23:45 in 15 minute intervals."
        )
      end
      change_setting(hold_until: time, hold: :temporary)
    end

    def output_status
      get_status
      status = @status['latestData']['uiData']['EquipmentOutputStatus']
      status = no_refresh { fan_running? ? 3 : status } if status.zero?
      EQUIPMENT_OUTPUT_STATUS[@status['latestData']['uiData']['EquipmentOutputStatus']]
    end

    def system_modes
      return @system_modes if @system_modes

      get_status if @status.empty?
      allowed_modes = [
        @status['latestData']['uiData']['SwitchEmergencyHeatAllowed'],
        @status['latestData']['uiData']['SwitchHeatAllowed'],
        @status['latestData']['uiData']['SwitchOffAllowed'],
        @status['latestData']['uiData']['SwitchCoolAllowed'],
        @status['latestData']['uiData']['SwitchAutoAllowed'],
      ]
      @system_modes = SYSTEM_MODES.select.with_index { |_, i| allowed_modes[i] }
    end

    def fan_modes
      return @fan_modes if @fan_modes

      get_status if @status.empty?
      allowed_modes = [
        @status['latestData']['fanData']['fanModeAutoAllowed'],
        @status['latestData']['fanData']['fanModeOnAllowed'],
        @status['latestData']['fanData']['fanModeCirculateAllowed'],
        @status['latestData']['fanData']['fanModeFollowScheduleAllowed']
      ]
      @fan_modes = FAN_MODES.select.with_index { |_, i| allowed_modes[i] }
    end

    def no_refresh(&block)
      @refresh = false
      result = yield
      @refresh = true
      result
    end

    private

    def get_status
      return @status unless @refresh || @status.empty?

      resp = @client.send(:request, "/Device/CheckDataSession/#{id}?_=#{Time.now.to_i}")
      @status = JSON.parse(resp.body)
    end

    # Required separation between high and low setpoints
    def deadband
      return @deadband if @deadband

      get_status if @status.empty?
      @deadband = @status['latestData']['uiData']['Deadband']
    end

    def min_cool_setpoint
      @info['ThermostatData']['MinCoolSetpoint']
    end

    def max_cool_setpoint
      @info['ThermostatData']['MaxCoolSetpoint']
    end

    def min_heat_setpoint
      @info['ThermostatData']['MinHeatSetpoint']
    end

    def max_heat_setpoint
      @info['ThermostatData']['MaxHeatSetpoint']
    end

    def raise_min_setpoint(min_temp, given_temp)
      raise TemperatureError.new("Minimum setpoint is #{min_temp}. Given: #{given_temp}")
    end

    def raise_max_setpoint(max_temp, given_temp)
      raise TemperatureError.new("Maximum setpoint is #{max_temp}. Given: #{given_temp}")
    end

    def payload(
      system_mode: nil,
      heat_setpoint: nil,
      cool_setpoint: nil,
      hold_until: nil,
      hold: nil,
      fan_mode: nil
    )
      {
        'DeviceID':       id,
        'SystemSwitch':   SYSTEM_MODES.index(system_mode),
        'HeatSetpoint':   heat_setpoint,  
        'CoolSetpoint':   cool_setpoint,
        'HeatNextPeriod': HOLD_TIMES.index(hold_until),
        'CoolNextPeriod': HOLD_TIMES.index(hold_until),
        'StatusHeat':     HOLD_TYPES.index(hold),
        'StatusCool':     HOLD_TYPES.index(hold),
        'FanMode':        FAN_MODES.index(fan_mode)
      }
    end

    def change_setting(**data)
      resp = @client.send(:request, '/Device/SubmitControlScreenChanges',
        method: 'post',
        data: payload(**data)
      )
      JSON.parse(resp.body)['success'] == 1
    end

    class FanError < StandardError; end
    class HoldError < StandardError; end
    class SystemError < StandardError; end
    class TemperatureError < StandardError; end
  end
end
